Extension/src/LanguageServer/Providers/callHierarchyProvider.ts

```diff
@@ -124,76 +130,54 @@ export class CallHierarchyProvider implements vscode.CallHierarchyProvider {
-                        this.logTelemetry(CallHierarchyCallsToEvent, requestStatus, progressBarDuration);
-                        reject(new vscode.CancellationError());
-                    } else {
-                        this.logTelemetry(CallHierarchyCallsToEvent, CallHierarchyRequestStatus.Succeeded, progressBarDuration);
-                        if (result?.calls.length === 0) {
-                            resolve(undefined);
-                        } else {
-                            resolve(this.createIncomingCalls(result.calls));
-                        }
-                    }
-
-                    this.client.clearPendingReferencesCancellations();
-                };
-
-                workspaceReferences.setCallHierarchyResultsCallback(resultCallback);
-                workspaceReferences.startCallHierarchyIncomingCalls(params);
-
-                token.onCancellationRequested(e => {
-                    if (params === DefaultClient.referencesParams) {
-                        this.client.cancelReferences();
-                    }
-                });
-            };
-
-            if (DefaultClient.referencesRequestPending || workspaceReferences.symbolSearchInProgress) {
-                const cancelling: boolean = DefaultClient.referencesPendingCancellations.length > 0;
-                DefaultClient.referencesPendingCancellations.push({
-                    reject: () => {
-                        // Complete with nothing instead of rejecting, to avoid an error message from VS Code
-                        resolve(undefined);
-                    }, callback
-                });
-                if (!cancelling) {
-                    workspaceReferences.referencesCanceled = true;
-                    this.client.languageClient.sendNotification(CancelReferencesNotification);
-                }
-            } else {
-                callback();
-            }
-        });
+        Promise<vscode.CallHierarchyIncomingCall[] | undefined> {
+        await this.client.awaitUntilLanguageClientReady();
+        workspaceReferences.cancelCurrentReferenceRequest(CancellationSender.NewRequest);
+
+        const CallHierarchyCallsToEvent: string = "CallHierarchyCallsTo";
+        if (item === undefined) {
+            this.logTelemetry(CallHierarchyCallsToEvent, CallHierarchyRequestStatus.Failed);
+            return undefined;
+        }
+
+        // Listen to a cancellation for this request. When this request is cancelled,
+        // use a local cancellation source to implicitly cancel a token.
+        let requestCanceled: CancellationSender | undefined;
+        const cancelSource: vscode.CancellationTokenSource = new vscode.CancellationTokenSource();
+        this.disposables.push(token.onCancellationRequested(() => {
+            requestCanceled = CancellationSender.ProviderToken;
+            cancelSource.cancel();
+        }));
+        this.disposables.push(workspaceReferences.onCancellationRequested(sender => {

```

Comment:
Rather than register a callback on this event every time a new request occurs (which will accumulate a ton of callbacks on that event, as these aren't currently being removed/disposed), maybe register just once for each provider, in the providers' constructors?  Then, if that used `this.cancelSource.cancel()`, it would always apply to the current in-progress operation.

Reply:
At the beginning of each new request, the event is called, won't the current request cancel it self if `this.cancelSource.cancel()` is used?

Reply:
I'm not sure I follow.  `workspaceReferences.cancelCurrentReferenceRequest(CancellationSender.NewRequest);` would get called (and call `this.cancelSource.,cancel()`) before you would replace `this.cancelSource` with a new one.